projectq/setups/trapped_ion_decomposer.py

```diff
@@ -0,0 +1,155 @@
+# If the value is -1 then the last gate applied (during a decomposition!) was Ry(-math.pi/2)
+# If the value is 1 then the last gate applied (during a decomposition!) was Ry(math.pi/2)
+# If the value is 0 then the last gate applied (during a decomposition!) was a Rx
+
+prev_Ry_sign=dict() # Keeps track of most recent Ry sign, i.e.
+                    # whether we had Ry(-pi/2) or Ry(pi/2)
+                    # prev_Ry_sign[qubit_index]  
+                    # should hold -1 or +1
+
+def chooser_Ry_reducer(cmd,decomposition_list):
+    """
+    Choose the decomposition so as to maximise Ry cancellations,
+    based on the previous decomposition used for the given qubit. 
+
+    Note:
+        Classical instructions gates e.g. Flush and 
+        Measure are automatically allowed.
+
+    Returns:
+        A decomposition object from the decomposition_list.
+    """
+    decomp_rule=dict() 
+    name='default'
+    
+    for el in decomposition_list:
+        try:
+            decomposition=el.decompose.__name__.split('_')
+            name=decomposition[2]
+            decomp_rule[decomposition[3]]=el
+            # 'M' stands for minus, 'P' stands for plus
+            # 'N' stands for neutral
+            # e.g. decomp_rule['M'] will give you the 
+            # decomposition_rule that ends with a Ry(-pi/2)
+        except:
+            pass
+    if name=='cnot2rxx': 
+        ctrl = cmd.control_qubits
+        idx=str(ctrl) # index of the qubit
+        if idx in prev_Ry_sign:
+            if prev_Ry_sign[idx]<=0:
+                # If the previous qubit had Ry(-pi/2)
+                # choose the decomposition that starts 
+                # with Ry(pi/2)
+                prev_Ry_sign[idx]=-1
+                # Now the prev_Ry_sign is set to -1
+                # since at the end of the decomposition
+                # we will have a Ry(-pi/2)
+                return decomp_rule['M']
+            else:
+                # Previous qubit had Ry(pi/2)
+                # choose decomposition that starts 
+                # with Ry(-pi/2) and ends with R(pi/2)
+                prev_Ry_sign[idx]=1
+                return decomp_rule['P']
+        else:
+            # In this case where we have no previous Ry sign 
+            # we choose decomp_rule['M']
+            prev_Ry_sign[idx]=-1
+            return decomp_rule['M']
+    elif name=='h2rx':
+        # Block operates similar to 'cnot2rxx' case
+        qubit = cmd.qubits[0]
+        idx=str(qubit)
+        if idx not in prev_Ry_sign:
+            prev_Ry_sign[idx]=+1
+            return decomp_rule['M']
+        elif prev_Ry_sign[idx]==0:
+            prev_Ry_sign[idx]=+1
+            return decomp_rule['M']
+        else:
+            prev_Ry_sign[idx]=00
+            return decomp_rule['N']
+    elif name=='rz2rx':
+        # Block operates similar to 'cnot2rxx' case
+        qubit = cmd.qubits[0]
+        idx=str(qubit)
```

Comment:
Rather than converting the list of qubits to a string, I would use the qubit identifiers:

```python
idx = [qb.id for qb in cmd.qubits[0]]
```